### PR TITLE
Feature/article meta

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by ghost** [comment]
@xiphiaz Coverage has decreased because of a resolve I have added and untested author changes which I am still working on. This is good to merge though, it has article meta changes in it.

File: api/app/Http/Controllers/ArticlePermalinkController.php [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 class ArticlePermalinkController extends ChildEntityController
 {
-    protected $relationName = 'permalinks';
+    protected $relationName = 'articlePermalink';
```
Comment (zakhenry) [inline api/app/Http/Controllers/ArticlePermalinkController.php]: The relation should be pluralised if it is a `many` relation

File: api/app/Http/Controllers/ArticleTagController.php [hunk 2]
```diff
@@ -26,7 +26,7 @@
 
 class ArticleTagController extends ChildEntityController
 {
-    protected $relationName = 'tags';
+    protected $relationName = 'tag';
```
Comment (zakhenry) [inline api/app/Http/Controllers/ArticleTagController.php]: Should be plural

File: app/src/app/admin/articles/article/article.ts [hunk 3]
```diff
@@ -78,12 +106,13 @@ namespace app.admin.articles.article {
...
-        static $inject = ['article', '$stateParams', 'articleService', 'notificationService'];
+        static $inject = ['article', '$stateParams', 'articleService', 'notificationService', 'articleMetaTags'];
         constructor(
             public article:common.models.Article,
             public $stateParams:IArticleStateParams,
             private articleService:common.services.article.ArticleService,
-            private notificationService:common.services.notification.NotificationService) {
+            private notificationService:common.services.notification.NotificationService
+        ) {
         }
```
Comment (zakhenry) [inline app/src/app/admin/articles/article/article.ts:101]: This controller isn't using articleMetaTags, so it shouldn't be injected - it is better to inject it into the controller that actually needs it.

File: app/src/app/admin/articles/article/article.ts [hunk 4]
```diff
@@ -61,6 +65,30 @@ namespace app.admin.articles.article {
...
+                            })
+                            .thru((templateMeta) => {
+                                let leftovers = _.filter(article._articleMeta, (metaTag) => {
+                                    return !_.contains(templateMeta, metaTag);
+                                });
+
+                                return templateMeta.concat(leftovers);
+                            })
+                            .value();
+
```
Comment (zakhenry) [inline app/src/app/admin/articles/article/article.ts:74]: This resolve should either
a) Return the meta tags so the meta controller can edit them directly (assuming the binding still works)
b) If the binding doesn't work, move the logic to the article resolve function (probably create a static method in the article service to limit the logic that is in the article resolve. eg `hydrateMetaFromTemplate(meta:common.models.ArticleMeta[], template:string[]):meta:common.models.ArticleMeta`

File: app/src/app/admin/articles/article/meta/meta.spec.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        });
+
+        afterEach(() => {
+
+        });
+
+
+        it('should run at least one test', () => {
+
+            expect(true).to.be.true;
```
Comment (zakhenry) [inline app/src/app/admin/articles/article/meta/meta.spec.ts]: Uh?

File: app/src/app/admin/articles/article/article.ts (hunk 3 above)
Comment (ghost) [inline app/src/app/admin/articles/article/article.ts:101]: @xiphiaz I thought you needed to inject it for it to perform the resolve, turns out I was wrong =O

File: app/src/app/admin/articles/article/article.ts (hunk 4 above)
Comment (ghost) [inline app/src/app/admin/articles/article/article.ts:74]: a) ~~Binding doesn't work.~~ Binding may or may not work, thinking about how I tested it before I may or may not have done it right. Have implemented b) so will leave as is; I think b) is cleaner anyways.
b) ~~Will implement this.~~ Have implemented this.

File: app/src/app/admin/articles/article/meta/meta.spec.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                MetaController = $controller(app.admin.articles.article.meta.namespace + '.controller', {
+                    article: article,
+                    notificationService: notificationService
+                });
+            });
+
+        });
+
+    });
+
```
Comment (zakhenry) [inline app/src/app/admin/articles/article/meta/meta.spec.ts:56]: This spec isn't testing anything?

**Comment by ghost** [comment]
Coverage rose to 95.61% from 95.33% (+0.28%).

**Comment by ghost** [comment]
Coverage same as before (95.61%).

**Pull request merged**